address/address.go

```diff
@@ -1,41 +1,71 @@
 package address
 
 import (
+	"bytes"
+	"encoding/json"
 	"fmt"
+	"strconv"
 	"strings"
 
-	"gx/ipfs/QmVmDhyTTUcQXFD1rRQ64fGLMSAoaQvNH3hwuaCFAPq2hy/errors"
+	"gx/ipfs/QmSKyB5faguXT4NqbrXpnRXqaVj5DhSm7x9BtzFydBY1UK/go-leb128"
+	errors "gx/ipfs/QmVmDhyTTUcQXFD1rRQ64fGLMSAoaQvNH3hwuaCFAPq2hy/errors"
 	"gx/ipfs/QmZp3eKdYQHHAneECmeK6HhiMwTPufmjC8DuuaGKv3unvx/blake2b-simd"
+	logging "gx/ipfs/QmbkT7eMTyXfpeyB3ZMxxcxg7XH8t6uXp49jqzz4HB7BGF/go-log"
+	cbor "gx/ipfs/QmcZLyosDwMKdB6NLRsiss9HXzDPhVhhRtPy67JFKTDQDX/go-ipld-cbor"
+	"gx/ipfs/QmdBzoMxsBpojBfN1cv5GnKtB7sfYBMoLH7p9qSyEVYXcu/refmt/obj/atlas"
 )
 
+var log = logging.Logger("address")
+
 func init() {
-	// cbor.RegisterCborType(Address{})
-}
+	cbor.RegisterCborType(addressAtlasEntry)
 
-var hashConfig = &blake2b.Config{Size: HashLength}
+	TestAddress = NewActorAddress([]byte("satoshi"))
+	TestAddress2 = NewActorAddress([]byte("nakamoto"))
 
-// Hash hashes the given input using the default address hashing function,
-// Blake2b 160.
-func Hash(input []byte) []byte {
-	hasher, err := blake2b.New(hashConfig)
-	if err != nil {
-		// If this happens sth is very wrong.
-		panic(fmt.Sprintf("invalid address hash configuration: %v", err))
-	}
-	if _, err := hasher.Write(input); err != nil {
-		// blake2bs Write implementation never returns an error in its current
-		// setup. So if this happens sth went very wrong.
-		panic(fmt.Sprintf("blake2b is unable to process hashes: %v", err))
-	}
-	return hasher.Sum(nil)
+	NetworkAddress = NewActorAddress([]byte("filecoin"))
+	StorageMarketAddress = NewActorAddress([]byte("storage"))
+	PaymentBrokerAddress = NewActorAddress([]byte("payments"))
 }
 
-// MakeTestAddress creates a new testnet address by hashing the given string
-func MakeTestAddress(input string) Address {
-	hash := Hash([]byte(input))
+var addressAtlasEntry = atlas.BuildEntry(Address{}).Transform().
+	TransformMarshal(atlas.MakeMarshalTransformFunc(
+		func(a Address) ([]byte, error) {
+			return []byte(a.str), nil
+		})).
+	TransformUnmarshal(atlas.MakeUnmarshalTransformFunc(
+		func(x []byte) (Address, error) {
+			return Address{string(x)}, nil
```

Comment:
This should be calling NewFromBytes, so that validation checks happen. Otherwise untrusted cbor input could give us potentially bad addresses.

Reply:
addressed